### DISCUSSION TITLE
Migrate from v5 to v6

### DISCUSSION THREAD

**Discussion by qetza** [qetza/replacetokens-task#34] [Announcements]
Most inputs where renamed, migrated or have their values changed, outputs were renamed and some transforms were renamed between v5 and v6.

When migrating from v5 to v6 use the following tables to migrate needed inputs, outputs and transforms:

**inputs**
| v5 | v6 | Comments |
|-|-|-|
| actionOnNoFiles:&nbsp;continue | ifNoFilesFound:&nbsp;ignore |  |
| actionOnNoFiles:&nbsp;fail | ifNoFilesFound:&nbsp;error |  |
| actionOnNoFiles:&nbsp;* | ifNoFilesFound:&nbsp;* | rename only for other values |
| defaultValue:&nbsp;* | missingVarDefault:&nbsp;* | rename only |
| emptyValue |  | remove input |
| enableRecursion:&nbsp;* | recursive:&nbsp;* | rename only |
| enableTelemetry:&nbsp;false | telemetryOptout:&nbsp;true |  |
| enableTelemetry:&nbsp;true | telemetryOptout:&nbsp;false |  |
| enableTransforms:&nbsp;* | transforms:&nbsp;* | rename only |
| encoding:&nbsp;win1252 | encoding:&nbsp;windows1252 |  |
| escapeType: none | escape: off |  |
| escapeType:&nbsp;* | escape:&nbsp;* | rename only for other values |
| inlineVariables:&nbsp;* | additionalVariables:&nbsp;* | set values as an object |
| keepToken:&nbsp;false</br>actionOnMissing:&nbsp;continue</br>useDefaultValue:&nbsp;false | missingVarAction:&nbsp;none</br>missingVarLog:&nbsp;off | silently replace missing tokens with an empty string |
| keepToken:&nbsp;false</br>actionOnMissing:&nbsp;continue</br>useDefaultValue:&nbsp;true | missingVarAction:&nbsp;replace | silently replace missing tokens with default value  |
| keepToken:&nbsp;false</br>actionOnMissing:&nbsp;warn</br>useDefaultValue:&nbsp;false | missingVarAction:&nbsp;none</br>missingVarLog:&nbsp;warn | replace missing tokens with an empty string and log a warning message |
| keepToken:&nbsp;false</br>actionOnMissing:&nbsp;warn</br>useDefaultValue:&nbsp;true | missingVarAction:&nbsp;replace</br>missingVarLog:&nbsp;warn | replace missing tokens with default value and log a warning message |
| keepToken:&nbsp;false</br>actionOnMissing:&nbsp;fail</br>useDefaultValue:&nbsp;false | missingVarAction:&nbsp;none</br>missingVarLog:&nbsp;error | replace missing tokens with an empty string and log an error message which will make the task fail |
| keepToken:&nbsp;false</br>actionOnMissing:&nbsp;fail</br>useDefaultValue:&nbsp;true | missingVarAction:&nbsp;replace</br>missingVarLog:&nbsp;error | replace missing tokens with default value and log an error message which will make the task fail |
| keepToken:&nbsp;true</br>actionOnMissing:&nbsp;continue | missingVarAction:&nbsp;keep</br>missingVarLog:&nbsp;off | keep missing tokens |
| keepToken:&nbsp;true</br>actionOnMissing:&nbsp;warn | missingVarAction:&nbsp;keep</br>missingVarLog:&nbsp;warn | keep missing tokens and log a warning message |
| keepToken:&nbsp;true</br>actionOnMissing:&nbsp;fail | missingVarAction:&nbsp;keep</br>missingVarLog:&nbsp;error | keep missing tokens and log an error message which will make the task fail |
| rootDirectory:&nbsp;* | root:&nbsp;* | rename only |
| targetFiles:&nbsp;* | sources:&nbsp;* | <ul><li>comma-separated values are not supported anymore, if you have multiple entries use newline to separate them</li><li>only slashes are supported as directory separator</li></ul> |
| tokenPattern:&nbsp;rm | tokenPattern:&nbsp;doubleunderscores |  |
| transformPrefix:&nbsp;* | transformsPrefix:&nbsp;* | rename only |
| transformSuffix:&nbsp;* | transformsSuffix:&nbsp;* | rename only |
| useLegacyEmptyFeature:&nbsp;true |  | not supported |
| useLegacyEmptyFeature:&nbsp;false |  | remove input |
| useLegacyPattern:&nbsp;true |  | not supported |
| useLegacyPattern:&nbsp;false |  | remove input |
| variableFiles:&nbsp;* | additionalVariables:&nbsp;* | set paths as an array of strings starting with `@` and use only slashes as directory separator |
| variableSeparator:&nbsp;* | separator:&nbsp;* | rename only |
| verbosity:&nbsp;detailed | logLevel:&nbsp;debug |  |
| verbosity:&nbsp;normal | logLevel:&nbsp;info |  |
| verbosity:&nbsp;off | logLevel:&nbsp;error |  |
| writeBOM:&nbsp;* | addBOM:&nbsp;* | rename and changed default value to `false`; if you didn't specify writeBOM, add addBOM:&nbsp;true |

**outputs**
| v5 | v6 |
|-|-|
| defaultValueCount | defaults |
| fileProcessedCount | files |
| tokenFoundCount | tokens |
| tokenReplacedCount | replaced |
| transformExecutedCount | transforms |

**transforms**
| v5 | v6 |
|-|-|
| noescape(*) | raw(*) |

**Comment by MattacksWTW** [comment]
To be proper English 'ifNoFilesFound' would have been better phrased 'whenNoFilesFound'
Etc.

**Reply by matthewc-dccu** [reply]
I believe `ifNoFilesFound` is correct English in this context.

- **If** is used to discuss conditional or uncertain events, indicating the possibility that an event may or may not happen. 
- **When** is used when you're certain that an event will happen, and it pertains to timing rather than conditionality.

**Comment by parshOne** [comment]
Could you please tell me how long then v5 will be around and supported for?

**Reply by qetza** [reply]
Hi @parshOne,
I don't intend to remove any version so all versions will still be available in the extension.

As for support i don't think the v5 version will be deprecated soon, i'll still update it to support new node version but for new features it will depend it will depend on the time needed to implement them as for new features i will privilege version 6 as it's using a common library which allows the extension to be available as a CLI and a GitHub Action.

In any case as the code is open-source anyone can fork it and update it as needed :smile:

**Comment by tarun-ghcp** [comment]
I am using Replace Token v6 in Azure DevOps pipeline with Angular 15 based application. The pipeline have a echo task that correctly displays the environment variables and their values just before the 'Replace Token' task. The filepath of main.xxxxxxxx.js is also correctly displayed in the Replace token logs. Also, the message says 'replaced 7 of 7 tokens' that means tokens are successfully replaced by the actual values. But the main.xxxxxxxxxx.js when checked contains the placeholders only. The format used in pipeline task is 'custom' with prefix and suffix as '#{' and '}#'. Below is how my environment.ts file looks like:

production: false,
datamanager_fn_url: "#{Env.DataManagerUrl}#",
sourcesystem: "#{Env.SourceSystem}#",
clientId: "#{Env.ClientID}#",
tenant: "#{Env.Tenant}#",
authority: "#{Env.Authority}#",
redirectUri: "#{Env.RedirectUri}#",
redirectUrl: "#{Env.RedirectUrl}#"

**Reply by tarun-ghcp** [reply]
Hey @qetza , can you just have a look please and suggest?

> I am using Replace Token v6 in Azure DevOps pipeline with Angular 15 based application. The pipeline have a echo task that correctly displays the environment variables and their values just before the 'Replace Token' task. The filepath of main.xxxxxxxx.js is also correctly displayed in the Replace token logs. Also, the message says 'replaced 7 of 7 tokens' that means tokens are successfully replaced by the actual values. But the main.xxxxxxxxxx.js when checked contains the placeholders only. The format used in pipeline task is 'custom' with prefix and suffix as '#{' and '}#'. Below is how my environment.ts file looks like:
> 
> production: false, datamanager_fn_url: "#{Env.DataManagerUrl}#", sourcesystem: "#{Env.SourceSystem}#", clientId: "#{Env.ClientID}#", tenant: "#{Env.Tenant}#", authority: "#{Env.Authority}#", redirectUri: "#{Env.RedirectUri}#", redirectUrl: "#{Env.RedirectUrl}#"

**Reply by qetza** [reply]
Hi @tarun-ghcp,
You can turn debug logging with the parameter `logLevel: debug`, this will display in the log every token found and the value it is replaced with.
Could you share also your task configuration?

**Reply by tarun-ghcp** [reply]
Hi @qetza, thanks for looking into it. The LogLevel already set to debug. 

Here is the YAML for the Replace task:

steps:
- task: qetza.replacetokens.replacetokens-task.replacetokens@6
  displayName: 'Replace environment variable tokens in **/main.*.js'
  inputs:
    root: '$(System.DefaultWorkingDirectory)/_Hatch.Bim.PMI-Dev-Private/AngularAppDropFiles'
    sources: '**/main.*.js'
    tokenPattern: custom
    tokenPrefix: '#{'
    tokenSuffix: '}#'
    addBOM: true
    logLevel: debug
    ifNoFilesFound: error
    
    
   Also please see the attached logs of Replace task:
<img width="1205" height="484" alt="Replace_token_logs" src="https://github.com/user-attachments/assets/0928740b-6be3-4343-9082-6b4f13aabb45" />

**Reply by qetza** [reply]
Hi @tarun-ghcp,

Based on the logs the task has found 7 tokens and replaced them with 7 values in the file `F:\devopsagent\_work\r1910\a\_Hatch.Bim.PMI-Dev-Private\AngularAppDropFiles\dist\main.12604b58da5fb....js`. I don't see any issue here.

Can you make sure that the file you are checking after the replace tokens is the one specified in the logs?

**Reply by tarun-ghcp** [reply]
Hi @qetza 
Yes i checked the same main.*****.js file as per the path in logs and found the placeholders. Even in some cases I reverted to version 3.x and found same outcome...placeholders not getting replaced.

**Reply by qetza** [reply]
Hi @tarun-ghcp,
That's strange as the logs indicates the replacements. Do you see anything if you enable system diagnostics?

**Comment by tarun-ghcp** [comment]
Hi @qetza , When we added the task in 'build' pipeline, it worked. So, for now, will use it in in build instead of 'release' pipeline.
<img width="1167" height="392" alt="image" src="https://github.com/user-attachments/assets/bb6313ef-7060-48a5-9b4c-cb2972703fbf" />
